File: src/org/jgll/parser/GSSEdge.java
package org.jgll.parser;

import org.jgll.sppf.SPPFNode;
import org.jgll.util.hashing.HashFunction;
import org.jgll.util.hashing.HashKey;
import org.jgll.util.hashing.Level;

/**
 * 
 * The values needed to uniquely identify a GSS edge are
 * the Source GSS node's grammar label, input index,
 * SPPF node's grammar label, and the Destination GSS node's
 * grammar label and input index.
 * 
 * Note that the SPPF's left extent is equal to the destination node's
 * input index and its right extent is equal to the source node's
 * input index. However, since Dummy nodes don't have input
 * indices, we use the input indices from GSS nodes to 
 * uniquely identify a GSS edge.
 * 
 * src (L1, j)
 * dst (L2, i)
 * node (L, i, j)
 * 
 * @author Ali Afroozeh
 * 
 */
public class GSSEdge implements Level, HashKey {

	private final GSSNode src;
	private final SPPFNode sppfNode;
	private final GSSNode dst;
	
	public GSSEdge(GSSNode src, SPPFNode sppfNode, GSSNode dst) {
		this.src = src;
		this.sppfNode = sppfNode;
		this.dst = dst;	
	}		
	
	public SPPFNode getSppfNode() {
		return sppfNode;
	}
	
	public GSSNode getDestination() {
		return dst;
	}
	
	@Override
	public int hashCode() {
		return HashFunctions.defaulFunction().hash(src.getGrammarSlot().getId(),
											       src.getInputIndex(),
												   dst.getGrammarSlot().getId(),
												   dst.getInputIndex(),
												   sppfNode.hashCode());
	}
	
	@Override
	public int hash(HashFunction f) {
		return f.hash(src.getGrammarSlot().getId(),
				      src.getInputIndex(),
					  dst.getGrammarSlot().getId(),
					  dst.getInputIndex(),
					  sppfNode.hash(f));
	}
	
	@Override
	public boolean equals(Object o) {
		
		if(this == o) {
			return true;
		}
		
		if(! (o instanceof GSSEdge)) {
			return false;
		}
		
		GSSEdge other = (GSSEdge) o;
		
		return src.getGrammarSlot() == other.src.getGrammarSlot() &&
			   src.getInputIndex() == other.src.getInputIndex() &&
			   dst.getGrammarSlot() == other.dst.getGrammarSlot() &&
			   dst.getInputIndex() == other.dst.getInputIndex() &&
			   sppfNode.equals(other.sppfNode);
	}

	@Override
	public int getLevel() {
		return src.getInputIndex();
	}
	
}

File: src-scala/org/jgll/parser/GSSEdge.scala
package org.jgll.parser

import org.jgll.sppf.{TerminalSymbolNodeTrait, SPPFNodeTrait}
import org.jgll.util.hashing.ExternalHasher
import org.jgll.util.hashing.Level
import org.jgll.util.hashing.hashfunction.HashFunction
import scala.reflect.BeanProperty

trait GSSEdgeTrait {
  self: SPPFNodeTrait
   with TerminalSymbolNodeTrait
   with GSSNodeTrait =>

  import GSSEdge._
  object GSSEdge {

    val externalHasher = new GSSEdgeExternalHasher()

    val levelBasedExternalHasher = new GSSEdgeExternalHasher()

    private def compare(first: SPPFNode, second: SPPFNode): Boolean = {
      if (first.isInstanceOf[TerminalSymbolNode] && second.isInstanceOf[TerminalSymbolNode]) {
        first.asInstanceOf[TerminalSymbolNode].getMatchedChar ==
          second.asInstanceOf[TerminalSymbolNode].getMatchedChar
      } else {
        first.getGrammarSlot == second.getGrammarSlot
      }
    }

    @SerialVersionUID(1L)
    class GSSEdgeExternalHasher extends ExternalHasher[GSSEdge] {

      override def hash(edge: GSSEdge, f: HashFunction): Int = {
        if (edge.sppfNode.isInstanceOf[TerminalSymbolNode]) {
          f.hash(Array(edge.src.getGrammarSlot.getId, edge.src.getInputIndex, edge.dst.getGrammarSlot.getId,
            edge.dst.getInputIndex, 31, edge.sppfNode.asInstanceOf[TerminalSymbolNode].getMatchedChar))
        } else {
          f.hash(Array(edge.src.getGrammarSlot.getId, edge.src.getInputIndex, edge.dst.getGrammarSlot.getId,
            edge.dst.getInputIndex, 17, edge.sppfNode.getGrammarSlot.getId))
        }
      }

      override def equals(e1: GSSEdge, e2: GSSEdge): Boolean = {
        e1.src.getGrammarSlot == e2.src.getGrammarSlot && e1.src.getInputIndex == e2.src.getInputIndex &&
          e1.dst.getGrammarSlot == e2.dst.getGrammarSlot &&
          e1.dst.getInputIndex == e2.dst.getInputIndex &&
          compare(e1.sppfNode, e2.sppfNode)
      }
    }

    @SerialVersionUID(1L)
    class GSSEdgeLevelBasedExternalHasher extends ExternalHasher[GSSEdge] {

      override def hash(edge: GSSEdge, f: HashFunction): Int = {
        if (edge.sppfNode.isInstanceOf[TerminalSymbolNode]) {
          f.hash(edge.src.getGrammarSlot.getId, edge.dst.getGrammarSlot.getId, edge.dst.getInputIndex,
            31, edge.sppfNode.asInstanceOf[TerminalSymbolNode].getMatchedChar)
        } else {
          f.hash(edge.src.getGrammarSlot.getId, edge.dst.getGrammarSlot.getId, edge.dst.getInputIndex,
            17, edge.sppfNode.getGrammarSlot.getId)
        }
      }

      def equals(e1: GSSEdge, e2: GSSEdge): Boolean = {
        e1.src.getGrammarSlot == e2.src.getGrammarSlot && e1.dst.getGrammarSlot == e2.dst.getGrammarSlot &&
          e1.dst.getInputIndex == e2.dst.getInputIndex &&
          compare(e1.sppfNode, e2.sppfNode)
      }
    }
  }

  class GSSEdge(private val src: GSSNode, @BeanProperty val sppfNode: SPPFNode, private val dst: GSSNode)
      extends Level {

    def getDestination(): GSSNode = dst

    override def hashCode(): Int = {
      externalHasher.hash(this, HashFunctions.defaulFunction())
    }

    override def equals(o: Any): Boolean = {
      if (this == o) {
        return true
      }
      if (!(o.isInstanceOf[GSSEdge])) {
        return false
      }
      val other = o.asInstanceOf[GSSEdge]
      src.getGrammarSlot == other.src.getGrammarSlot && src.getInputIndex == other.src.getInputIndex &&
        dst.getGrammarSlot == other.dst.getGrammarSlot &&
        dst.getInputIndex == other.dst.getInputIndex &&
        compare(sppfNode, other.sppfNode)
    }

    override def getLevel(): Int = src.getInputIndex
  }
}